File: Grunt/Grunt/Models/HaloInfinite/MatchInfo.cs
// <copyright file="MatchInfo.cs" company="Den Delimarsky">
// Developed by Den Delimarsky.
// Den Delimarsky licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
// The underlying API powering Grunt is managed by 343 Industries and Microsoft. This wrapper is not endorsed by 343 Industries or Microsoft.
// </copyright>

using System;

namespace OpenSpartan.Grunt.Models.HaloInfinite
{
    /// <summary>
    /// Container for general match information.
    /// </summary>
    [IsAutomaticallySerializable]
    public class MatchInfo
    {
        /// <summary>
        /// Gets or sets the match start time.
        /// </summary>
        public DateTime? StartTime { get; set; }

        /// <summary>
        /// Gets or sets the match end time.
        /// </summary>
        public DateTime? EndTime { get; set; }

        /// <summary>
        /// Gets or sets the match duration.
        /// </summary>
        public TimeSpan Duration { get; set; }

        /// <summary>
        /// Gets or sets the match lifecycle mode.
        /// </summary>
        public LifecycleMode LifecycleMode { get; set; }

        /// <summary>
        /// Gets or sets the match game variant category.
        /// </summary>
        public GameVariantCategory? GameVariantCategory { get; set; }

        /// <summary>
        /// Gets or sets the match level ID.
        /// </summary>
        public Guid LevelId { get; set; }

        /// <summary>
        /// Gets or sets the map variant for the match.
        /// </summary>
        public GenericAsset? MapVariant { get; set; }

        /// <summary>
        /// Gets or sets the game variant for the match.
        /// </summary>
        public UGCGameVariant? UgcGameVariant { get; set; }

        /// <summary>
        /// Gets or sets the clearance ID.
        /// </summary>
        public Guid ClearanceId { get; set; }

        /// <summary>
        /// Gets or sets the associated playlist.
        /// </summary>
        public GenericAsset? Playlist { get; set; }

        /// <summary>
        /// Gets or sets the associated playlist experience.
        /// </summary>
        public PlaylistExperience? PlaylistExperience { get; set; }

        /// <summary>
        /// Gets or sets the playlist map-mode pair.
        /// </summary>
        public GenericAsset? PlaylistMapModePair { get; set; }

        /// <summary>
        /// Gets or sets the season ID.
        /// </summary>
        public string? SeasonId { get; set; }

        /// <summary>
        /// Gets or sets the playable duration.
        /// </summary>
        public TimeSpan PlayableDuration { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether teams are enabled.
        /// </summary>
        public bool? TeamsEnabled { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether team scoring is enabled.
        /// </summary>
        public bool? TeamScoringEnabled { get; set; }
    }
}

File: src/node/src/models/halo-infinite/match/match-info.ts
import type { LifecycleMode } from '../enums/lifecycle-mode';

/**
 * Generic asset reference used across multiple API responses.
 */
export interface GenericAsset {
  /** Unique asset identifier */
  AssetId?: string;
  /** Version identifier */
  VersionId?: string;
  /** Combined asset version identifier */
  AssetVersionId?: string;
  /** Display name */
  PublicName?: string;
}

/**
 * UGC game variant information in match context.
 */
export interface UgcGameVariant {
  /** Asset identifier */
  AssetId?: string;
  /** Version identifier */
  VersionId?: string;
  /** Display name */
  PublicName?: string;
}

/**
 * Playlist experience tracking.
 */
export interface PlaylistExperience {
  /** Experience value */
  Value?: number;
}

/**
 * Gameplay interaction type.
 */
export interface GameplayInteraction {
  /** Interaction type identifier */
  Value?: number;
}

/**
 * General information about a match.
 *
 * Contains metadata like timing, game mode, map, and playlist info.
 */
export interface MatchInfo {
  /** Match start time (ISO 8601) */
  StartTime?: string;
  /** Match end time (ISO 8601) */
  EndTime?: string;
  /** Match duration as ISO 8601 duration string (e.g., "PT10M30S") */
  Duration?: string;
  /** Lifecycle mode (matchmade, custom, local) */
  LifecycleMode?: LifecycleMode;
  /** Game variant category */
  GameVariantCategory?: number;
  /** Map/level identifier */
  LevelId?: string;
  /** Map variant information */
  MapVariant?: GenericAsset;
  /** UGC game variant (for custom games) */
  UgcGameVariant?: UgcGameVariant;
  /** Clearance ID used for the match */
  ClearanceId?: string;
  /** Playlist information */
  Playlist?: GenericAsset;
  /** Playlist experience info */
  PlaylistExperience?: PlaylistExperience;
  /** Map-mode pair info */
  PlaylistMapModePair?: GenericAsset;
  /** Season identifier */
  SeasonId?: string;
  /** Playable duration */
  PlayableDuration?: string;
  /** Whether teams were enabled */
  TeamsEnabled?: boolean;
  /** Whether team scoring was enabled */
  TeamScoringEnabled?: boolean;
  /** Gameplay interaction type */
  GameplayInteraction?: GameplayInteraction;
}
